### PR TITLE
skip moe_act recompute path when grad is disabled

### PR DESCRIPTION
**Pull request by raghulasokan**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
This PR helps fix a growing `TE queue(_is_first_fp8_module)` process state accumulation issue which occurs during MoE training with activation checkpointing enabled. The fix directly improves the overall step time increase pattern that is observed otherwise over longer MoE training runs(>10K steps).

## Description
During training with `moe_act` selective activation recompute, `TEGroupedMLP.forward()` runs two forward passes: the original forward pass (batch inputs through the network) and a recompute forward pass triggered during backward via a registered hook.

**The bug:** the original forward pass was always taking the `CheckpointWithoutOutput` path regardless of whether grad was enabled, leading to an entry push into `_is_first_fp8_module` TE queue, but never gets popped because there is no hook registered.

The fix guards the `CheckpointWithoutOutput` setup behind `torch.is_grad_enabled()`, so the original forward pass under `torch.no_grad()` takes the plain `else` branch, ensuring no dangling entries in the queue.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

**Review by cuichenx — APPROVED** [review]
LGTM, please open more PRs for other places in code wher activations recompute is used

File: megatron/core/transformer/moe/experts.py [hunk 1]
```diff
@@ -1022,7 +1022,10 @@ def glu(x):
                 intermediate_parallel = intermediate_parallel.to(original_dtype)
             return intermediate_parallel
 
-        if self.activation_recompute:
+        use_activation_recompute = (
+                self.activation_recompute and torch.is_grad_enabled()
```
Comment (wujingyue) [inline megatron/core/transformer/moe/experts.py:1026]: Drive-by comments: I'm not sure it's the right fix.

We tried to support activation recomputation in MFSDP and converged to https://github.com/NVIDIA/Megatron-LM/blob/3134ff9876d38965bbed0e5344c932c1f4fbe6cc/megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py#L36. Note that torch.utils.checkpoint has reentrant mode (default) and non-reentrant mode, which behave differently.

**Review by yaox12 — APPROVED** [review]

File: megatron/core/transformer/moe/experts.py (hunk 1 above)
Comment (raghulasokan) [reply megatron/core/transformer/moe/experts.py:1026]: @wujingyue 

if only `"moe_act"` is set(and not `"moe"`) in `recompute_modules`, then when `TEGroupedMLP` forward runs, it will have grad enabled, and `CheckpointWithoutOutput` needs to be invoked -> which cannot be done with this check `torch._C._current_graph_task_id()` because the autograd graph traversal has not yet started.

**Comment by wujingyue** [comment]
@raghulasokan, can you fix DCO?

**Review by yashaswikarnati — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

File: megatron/core/transformer/moe/experts.py (hunk 1 above)
Comment (wujingyue) [reply megatron/core/transformer/moe/experts.py:1026]: > because the autograd graph traversal has not yet started.

Are you sure about this? IIUC, autograd starts with `loss.backward()`, which covers the life of activation recomputation. FWIW, we didn't choose to use `no_grad` because one could train a model with parts of it under `no_grad`.

Comment (wujingyue) [inline megatron/core/transformer/moe/experts.py:1026]: We'd better add a code comment explaining **why**.

File: megatron/core/transformer/moe/experts.py [hunk 2]
```diff
@@ -1022,7 +1022,10 @@ def glu(x):
                 intermediate_parallel = intermediate_parallel.to(original_dtype)
             return intermediate_parallel
 
-        if self.activation_recompute:
+        use_activation_recompute = (
```
Comment (wujingyue) [inline megatron/core/transformer/moe/experts.py:1025]: Consider renaming. `use_activation_recompute` as a name is confusing. It's not much different from `activation_recompute` and doesn't reflect `grad_enabled`.

File: megatron/core/transformer/moe/experts.py (hunk 1 above)
Comment (wujingyue) [inline megatron/core/transformer/moe/experts.py:1026]: Could we add a unit test? It's easy to forget about having to check `grad_enabled` and cause regression.

**Review by wujingyue — APPROVED** [review]

**Comment by asolergi-nv** [comment]
@raghulasokan Can you please address @wujingyue comments? Thanks!

File: megatron/core/transformer/moe/experts.py (hunk 1 above)
Comment (raghulasokan) [reply megatron/core/transformer/moe/experts.py:1061]: comment is added

File: megatron/core/transformer/moe/experts.py (hunk 2 above)
Comment (raghulasokan) [reply megatron/core/transformer/moe/experts.py]: have modified it to setup_activation_checkpoint as the flag's main purpose is to use moe_act checkpointing or not. the flag's state now depends on grad_enabled which is now specified as part of the code comment.

**Comment by Connor-XY** [comment]
@raghulasokan Thanks for addressing the code-comment and flag-name feedback. The [unit-test request](https://github.com/NVIDIA/Megatron-LM/pull/7030#discussion_r4009133609) is still unanswered, and the current diff changes only `experts.py`. Please add a focused regression covering the no-grad selective `moe_act` path and the normal grad-enabled path, and reply to the [remaining recomputation-mode question](https://github.com/NVIDIA/Megatron-LM/pull/7030#discussion_r3907615932). That will make the remaining review concrete without revisiting the items already addressed.

**Comment by Connor-XY** [comment]
/ok to test [0cee75c](https://github.com/NVIDIA/Megatron-LM/pull/7030/commits/0cee75c58b80ddd43d5d038daca21c33b84657da)
